File: leetcode/algo2/day10/40_combination_sum_2.cpp
#include <vector>
#include <iostream>
#include <algorithm>
std::vector<std::vector<int>> res;

void search(std::vector<int> &candidates, int target, int pointer, std::vector<int> temp_arr, int temp_sum)
{
    if (temp_sum == target)
        res.push_back(temp_arr);
    if (temp_sum > target)
        return;
    int prev = -1;
    for (int i = pointer; i < candidates.size(); i++)
    {
        if (candidates[i] == prev)
            continue;
        temp_arr.push_back(candidates[i]);
        search(candidates, target, i + 1, temp_arr, temp_sum + candidates[i]);
        temp_arr.pop_back();
        prev = candidates[i];
    }
}

std::vector<std::vector<int>> combinationSum2(std::vector<int> &candidates, int target)
{
    std::sort(candidates.begin(), candidates.end());
    std::vector<int> temp_arr;
    int temp_sum = 0;
    int pointer = 0;
    search(candidates, target, pointer, temp_arr, temp_sum);
    return res;
}
int main()
{
    std::vector<int> nums = {10, 1, 2, 7, 6, 1, 5, 8}; // 1,1,2,5,6,7,10
    int target = 8;
    for (auto x : combinationSum2(nums, target))
    {
        for (auto y : x)
            std::cout << y << " ";
        std::cout << std::endl;
    }
}

File: leetcode/algo2/day10/40_combination_sum_2.js
let res = [];

var solve = (nums, target, curr, pointer, temp_sum) => {
    if (temp_sum === target) res.push([...curr]);
    if (temp_sum > target) return;
    let prev = -1;
    for (let i = pointer; i < nums.length; i++) {
        if (nums[i] === prev) continue;
        curr.push(nums[i]);
        solve(nums, target, curr, i + 1, temp_sum + nums[i]);
        curr.pop();
        prev = nums[i];
    }
}


var combinationSum2 = function (nums, target) {
    nums.sort((a, b) => a - b);
    solve(nums, target, [], 0, 0);
    return res;
};
let nums = [2, 5, 2, 1, 2];
let target = 5;
for (const x of combinationSum2(nums, target)) {
    console.log(x);
}